services/provider_credential_service.rb

```diff
@@ -0,0 +1,59 @@
+require_relative "./services"
+require_relative "../shared/logging_module"
+
+module FastlaneCI
+  # Provides access to provider credential related methods
+  class ProviderCredentialService
+    include FastlaneCI::Logging
+
+    # Creates a new provider credential, and adds it to the User's provider
+    # credentials array
+    def create_provider_credential!(
+      user_id: nil, id: nil, email: nil, api_token: nil, provider_name: nil,
+      full_name: nil, type: nil
+    )
+      provider_credential = GitHubProviderCredential.new(
+        id: id, email: email, api_token: api_token, provider_name: provider_name,
+        full_name: full_name, type: type
+      )
+      user = Services.user_service.find_user(id: user_id)
+
+      if user.nil?
+        logger.error("Can't create provider credential for user, since user does not exist.")
+      else
+        new_user = User.new(
+          id: user.id,
+          email: user.email,
+          password_hash: user.password_hash,
+          provider_credentials: user.provider_credentials.push(provider_credential)
+        )
+        Services.user_service.update_user!(user: new_user)
+      end
+    end
+
+    def update_provider_credential!(
```

Comment:
This is trickier than it seems...Provider credentials can have different email addresses from the user it is associated with. 
For example: I might use `taquitos@google.com` as my fastlane.ci account, but I want to use `taquitos@gmail.com` for my GitHub credentials.
This is why we have a back reference to `credential.ci_user`. I think the better move is to lookup the user you want to add a credential to, then create the credential with the proper credential bits, and then save that back.

Reply:
I'm doing the user lookup in the method by [`user_id`](https://github.com/fastlane/ci/pull/221/files/37ab9934a7fcfa659deeda4d3bbc5064140d05e8#diff-a47cf038d988ee52e0479961109c6133R42) instead of by `email`. In the `update` view form for provider credentials, the `user_id` associated with the credential is [passed into the form](https://github.com/fastlane/ci/pull/221/files/37ab9934a7fcfa659deeda4d3bbc5064140d05e8#diff-c4965ee0be82e539c898652971f42e28R8).

Reply:
Ahh that makes sense. Carry on 😄